skfp/utils/splitters.py

```diff
@@ -0,0 +1,479 @@
+    """
+    Fill in missing sizes for train and test sets based on the provided sizes.
+    """
+    if train_size is None and test_size is None:
+        train_size = 0.8
+        test_size = 0.2
+    if train_size is None:
+        if test_size is not None:
+            train_size = 1 - test_size
+        else:
+            raise ValueError("test_size must be provided when train_size is None")
+    elif test_size is None:
+        test_size = 1 - train_size
+    return train_size, test_size
+
+
+def _get_data_from_indices(
+    data: Sequence[Union[str, Mol]], indices: Sequence[int]
+) -> list[Union[str, Mol]]:
+    """
+    Helper function to retrieve data elements from specified indices.
+    """
+    indices = set(indices)
+    return [data[idx] for idx in indices]
+
+
+def _split_additional_data(
+    additional_data: list[Sequence[Any]], *indices_lists: list[list[int]]
+) -> list[Sequence[Any]]:
+    """
+    Split additional data based on indices lists.
+    """
+    return list(
+        chain.from_iterable(
+            (_safe_indexing(a, indices),)
+            for a in additional_data
+            for indices in indices_lists
+        )
+    )
+
+
+def _split_ids(
+    scaffold_sets: list[list[int]], total_data_len: int, test_size: float
+) -> tuple[list[int], list[int]]:
+    """
+    Split IDs into training and testing sets based on scaffold sets.
+    """
+    train_ids: list[int] = []
+    test_ids: list[int] = []
+    desired_test_size = int(test_size * total_data_len)
+
+    for scaffold_set in scaffold_sets:
+        if len(test_ids) < desired_test_size:
+            test_ids.extend(scaffold_set)
+        else:
+            train_ids.extend(scaffold_set)
+
+    return train_ids, test_ids
+
+
+def _split_ids_three_sets(
+    scaffold_sets: list[list[int]],
+    total_data_len: int,
+    test_size: float,
+    valid_size: float,
+) -> tuple[list[int], list[int], list[int]]:
+    """
+    Split IDs into training, validation, and testing sets based on scaffold sets.
+    """
+    train_ids: list[int] = []
+    valid_ids: list[int] = []
+    test_ids: list[int] = []
+    desired_test_size = int(test_size * total_data_len)
+    desired_valid_size = int((test_size + valid_size) * total_data_len)
+
+    for scaffold_set in scaffold_sets:
+        if len(test_ids) < desired_test_size:
+            test_ids.extend(scaffold_set)
+        elif len(valid_ids) < desired_valid_size:
+            valid_ids.extend(scaffold_set)
+        else:
+            train_ids.extend(scaffold_set)
+
+    return train_ids, valid_ids, test_ids
+
+
+def _split_size(
+    train_size: Optional[float],
+    valid_size: Optional[float],
+    test_size: Optional[float],
+) -> tuple[float, float, float]:
+    """
+    Ensure the sum of train_size, valid_size, and test_size equals 1.0 and provide default values if necessary.
+    """
+    train_size, test_size, valid_size = _calculate_missing_sizes(
+        train_size, test_size, valid_size
+    )
```

Comment:
Paste that function here. We generally want to avoid functions used only in one place